dex/src/main/scala/com/wavesplatform/dex/SpendableBalancesActor.scala

```diff
@@ -0,0 +1,89 @@
+package com.wavesplatform.dex
+
+import akka.actor.{Actor, ActorRef}
+import akka.pattern.pipe
+import cats.instances.long.catsKernelStdGroupForLong
+import cats.syntax.group.catsSyntaxGroup
+import com.wavesplatform.dex.domain.account.Address
+import com.wavesplatform.dex.domain.asset.Asset
+import com.wavesplatform.dex.fp.MapImplicits.cleaningGroup
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class SpendableBalancesActor(spendableBalances: (Address, Set[Asset]) => Future[Map[Asset, Long]],
+                             allAssetsSpendableBalances: Address => Future[Map[Asset, Long]],
+                             addressDirectory: ActorRef)(implicit ec: ExecutionContext)
+    extends Actor {
+
+  var fullState: Map[Address, Map[Asset, Long]]  = Map.empty
+  var addressDff: Map[Address, Map[Asset, Long]] = Map.empty
+
+  def receive: Receive = {
+
+    case SpendableBalancesActor.Query.GetState(address, assets) =>
+      val maybeAddressState            = fullState.get(address).orElse(addressDff get address)
+      val assetsMaybeBalances          = assets.map(asset => asset -> maybeAddressState.flatMap(_ get asset)).toMap
+      val (knownAssets, unknownAssets) = assetsMaybeBalances.partition { case (_, balance) => balance.isDefined }
+      val knownPreparedState           = knownAssets.collect { case (a, Some(b)) => a -> b }
+
+      if (unknownAssets.isEmpty) sender ! SpendableBalancesActor.Reply.GetState(knownPreparedState)
+      else {
+        spendableBalances(address, unknownAssets.keySet).map { stateFromNode =>
+          if (fullState contains address) fullState = fullState.updated(address, fullState.getOrElse(address, Map.empty) ++ stateFromNode)
+          else addressDff = addressDff.updated(address, addressDff.getOrElse(address, Map.empty) ++ stateFromNode)
```

Comment:
We can't modify a state of the actor inside another thread. This could be solved through sending a message to `self`.

Reply:
Done